### PR TITLE
Refactoring of OpenCL profiling

### PR DESCRIPTION
**Pull request by jvillarre**
This pull request is a major change to how OpenCL level profiling is implemented, and separates out the profile summary generation/counter information from the trace generation into separate plugins.  The old architecture libraries are no longer built or used, although the code remains as a reference at this point.  This pull request also introduces new xrt.ini switches to independently control each aspect of profiling and deprecates some of the old switches (which are still supported at this time).

As part of this submission, the xocl to xdp side interface was changed, breaking the dependency between the xdp library and the internal xocl structures.  All callbacks to the plugins now transfer C objects and all xocl data structure manipulation is done on the xocl side (in the xocl/api/plugin/xdp directory).  The actual callback points in the xocl library have not changed, although some of the function signatures are different.

### CODE REVIEW

File: src/runtime_src/core/common/device.cpp [hunk 1]
```diff
@@ -92,8 +92,7 @@ get_xclbin_uuid() const
   catch (const query::no_such_key&) {
   }
 
-  // Emulation mode likely, just return m_xclbin_uuid which reflects
-  // the uuid of the xclbin loaded by this process.
+  // Emulation mode likely
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:95]: Is this reverting a comment that was recently added.   I'd like to preserve the comment.

File: src/runtime_src/xdp/profile/database/database.h [hunk 2]
```diff
@@ -40,7 +40,7 @@ namespace xdp {
   {
   public:
     // For messages sent to specific plugins
-    enum MessageType { } ;
+    enum MessageType { READ_COUNTERS } ;
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/database/database.h:43]: enum **class** with lower case enumeration values is preferred.

**Review by stsoe — COMMENTED** [review]
Looks good to me.  Just a request to preserve one comment.

File: src/runtime_src/core/common/device.cpp (hunk 1 above)
Comment (jvillarre) [reply src/runtime_src/core/common/device.cpp]: Restored the comment to the same as master in the latest commits.

**Review by stsoe — APPROVED** [review]

**Review by IshitaGhosh — APPROVED** [review]
The support for OpenCL profiling using new architecture in XDP looks good.

**Comment by dayeh-xilinx** [comment]
retest this please - after failing spooling of results for u30 due to python upgrade

**Comment by xilinxgitops** [comment]
retest this please

**Pull request merged**